Tar: Adjust the way we write GNU longlink and longpath metadata

carlossanlop opened the pull request:
The Tar entry types Longlink and Longpath, which are exclusive to the GNU format, are metadata entries used to store longer-than-expected link or name strings (respectively). They precede their actual entry, and they are not visible to the user, only the data they hold is visible.

Most of the regular header metadata in this metadata entries is not relevant, but we need to make sure it is well formed so other tools can read them properly.

There are 3 fields that were not getting their values stored as expected by other tools:

- The mtime, ctime and atime fields need to be set to the Unix Epoch so it shows up as zeros (0x30) by default. We were storing them as null (0x0) characters instead. For consistency, if a regular entry happens to get its mtime set to the Epoch, it should also be shown as zeros (0x30).
- The uid, gid, uname and gname fields were not being set. These fields aren't really important in a metadata entry, but other tools set this to the default value (uid=gid=0 and uname=gname=root). Based on the fact that 'a value' is preferred for these fields, we can set them to the same value that the real entry has (not root).
- The long text value (either link or name) should be null terminated, and this extra character should be counted in the size of the longpath. The library libdpkg, used for packagin in Debian, can handle a missing null terminator in longlink and longpath data starting with Debian 13, but we use Debian 12, and in that version, without the terminator, these entries are considered malformed.

carlossanlop committed:
Fix 3 GNU longlink and longpath bugs

The Tar entry types Longlink and Longpath, which are exclusive to the GNU format, are metadata entries used to store longer-than-expected link or name strings (respectively). They precede their actual entry, and they are not visible to the user, only the data they hold is visible.

Most of the regular header metadata in this metadata entries is not relevant, but we need to make sure it is well formed so other tools can read them properly.

There are 3 fields that were not getting their values stored as expected by other tools:

- The mtime, ctime and atime fields need to be set to the Unix Epoch so it shows up as zeros (0x30) by default. We were storing them as null (0x0) characters instead. For consistency, if a regular entry happens to get its mtime set to the Epoch, it should also be shown as zeros (0x30).
- The uid, gid, uname and gname fields were not being set. These fields aren't really important in a metadata entry, but other tools set this to the default value (uid=gid=0 and uname=gname=root). Based on the fact that 'a value' is preferred for these fields, we can set them to the same value that the real entry has (not root).
- The long text value (either link or name) should be null terminated, and this extra character should be counted in the size of the longpath. The library libdpkg, used by the dpkg-deb tool in Debian, now expects a null terminator in longlink and longpath data starting in Debian 13. Without the terminator, these entries are considered malformed.

carlossanlop committed:
Test to check mtime null values when set to UnixEpoch.

carlossanlop commented:
/backport to release/10.0-preview4

jkoritzinsky commented:
Slight adjustment to the statement about Debian: dpkg in Debian 12 and older expects the null terminator. Dpkg in Debian 13 can function safely without the null terminator.

carlossanlop commented:
> Slight adjustment to the statement about Debian: dpkg in Debian 12 and older expects the null terminator. Dpkg in Debian 13 can function safely without the null terminator.

Ah thanks for the clarification. I understood the opposite. I read the chat conversation again.

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -354,26 +354,33 @@ internal async Task WriteAsPaxAsync(Stream archiveStream, Memory<byte> buffer, C
+
+        // Checks if the name string is too long to fit in the regular header field.
+        // .NET strings do not include a null terminator by default, need to add it manually and also consider it for the length.
+        private bool IsNameTooLongForRegularField() => (Encoding.UTF8.GetByteCount(_name) + 1) > FieldLengths.Name;
```
_name is non nullable

carlossanlop committed:
Make sure empty uid and gid are written as '0' chars.

carlossanlop committed:
Adapt existing tests that check atime and ctime so that they also check those values in the LongLink and LongPath stream bytes themselves.
Also should check uid=gid=0 and the copying of uname and gname into the LongLink/LongPath entries.
Also should confirm that the data length in LongLink and LongPath entries is counting a null terminator in the data.

carlossanlop commented:
Shouldn't the bot copy commits to the backport PR which where pushed after the /backport command?

carlossanlop commented:
/backport to release/10.0-preview4

akoeplinger commented:
> Shouldn't the bot copy commits to the backport PR which where pushed after the /backport command?

no it needs another comment trigger like you did

tmds commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -623,18 +682,34 @@ private int WriteCommonFields(Span<byte> buffer, TarEntryType actualEntryType)
+            if (_format is TarEntryFormat.Gnu)
+            {
+                // Just like atime and ctime, mtime should be stored as zeros when the value is UnixEpoch.
+                checksum += WriteAsGnuTimestamp(_mTime, buffer.Slice(FieldLocations.MTime, FieldLengths.MTime));
```
I don't understand why we need a specific method `WriteAsGnuTimestamp`. It seems to handle the `DateTimeOffset.UnixEpoch` separate from any other timestamp?

I don't think such handling is necessary and that we should be able to remove that method and call `WriteAsTimestamp` directly.

carlossanlop replied:
I know `WriteAsTimestamp` has an `else` branch to check for Gnu, but when debugging I noticed we were not entering it for this case. I didn't want to mess with that code at the moment because even the slightest change was making several unrelated unit tests fail, so I decided to create a separate method for the Gnu timestamps.

See my main answer, I'd welcome addressing this better in a separate PR.

carlossanlop replied:
Method removed. We can keep using WriteAsTimestamp.

tmds commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -623,18 +682,34 @@ private int WriteCommonFields(Span<byte> buffer, TarEntryType actualEntryType)
+            else if (_uid == 0 && _format is TarEntryFormat.Gnu)
+            {
+                FormatAsZeroChars(buffer.Slice(FieldLocations.Uid, FieldLengths.Uid));
+            }
```
> The semantics of the code are: if the uid is less or equal to zero then the value in the buffer should represent zero.

It seems these are not the semantics, instead the semantics are: we're writing an invalid value in the buffer when the uid is not valid.

A uid of zero is the root user which should be an accepted value.

So the code should be:

```cs
if (_uid >= 0)
{
    checksum += FormatNumeric(_uid, buffer.Slice(FieldLocations.Uid, FieldLengths.Uid));
}
```

Same comment applies to gid.

carlossanlop replied:
I am aware that 0 is a valid uid/gid, but see my comment below, this is what the Debian tool is deciding to do with them (literal `0` chars instead of null chars). I'm just matching that.

carlossanlop commented:
@tmds here's why we need these changes:

![image](https://github.com/user-attachments/assets/664db43b-2a11-4a2d-b54c-8b2bcb4ff528)

This screenshot compares two tar files. On the left side is a tar file generated by Debian tooling. On the right side is the same file after we modify it with TarWriter to sign the binaries. The file on the right is the one that the Unified Build Debian tooling is telling us is malformed.

The error shows up when the tool attempts to analyze the entry for System.Runtime.InteropServices.RuntimeInformation.dll. This file has a large enough name that it does not fit inside the regular `name` field in the header, so a `LongLink` entry is needed to store that `name` in its data section.

Notice at the top you have the first 100 bytes which belong to the reserved name `././@LongLink`. Then you have the mode `0000644\0`.

And then you have, on the left, a uid and a gid of `0`, but represented as literaly `0x30` characters: `0000000\0`. We don't do that on the right side, so this PR takes care of that by reusing the uid and gid of the main entry.

Then you have the size of `00000000160\0` on the left and `00000000157\0` on the right. This is the main bug: we were not considering a null character as part of the text in the data section, so on the left side we have 111 bytes, but 110 bytes on the right (in octal both).

Then you have `mtime` set to `00000000000\0` on the left, but has mangled characters on the right because by default we were storing `DateTimeOffset.MinValue` in LongLing and LongPath entries. The fix for this is to store `DateTimeOffset.UnixEpoch`.

Then you have 8 bytes of checksum, 1 byte of typeflag `L`, then 100 bytes of link (empty), then 8 bytes of the magic string `ustar  `.

Then you have 8 bytes of `uname` and 8 bytes of `gname`, both which are empty on the right side. I am not sure why the original tool sets it to root, but the real entry's value should be considered a valid value to use, so we are fixing it by copying that value. Same as with the uid and gid above.

Then you have 8 bytes of devmajor and 8 bytes of devminor.

And finally we have the 12 bytes of atime and 12 bytes of mtime, also mangled due to the usage of `DateTimeOffset.MinValue`, which we're fixing with `DateTimeOffset.UnixEpoch`. This is the other big issue, we noticed with a previous PR where the _main entries_ were showing up as mangled due to invalid `atime` and `ctime` values. Looks to be the same problem but with LongLink and LongPath entries.

As to the way I fixed these, I think the code can always be improved, and I would welcome a separate discussion to do so. At the moment, our main focus is to unblock Unified Build, and these changes do that.

tmds commented:
@carlossanlop the changes I've requested are what the spec describes. I would not deviate from that unless it fixes a bug.

Let's look at what GNU `tar` does:

```
$ ls -n file
-rw-r--r--. 1 0 1000 0 Jan  1  1970 file
$ tar -cf archive.tar file
$ hexdump -C archive.tar 
00000000  66 69 6c 65 00 00 00 00  00 00 00 00 00 00 00 00  |file............|
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000060  00 00 00 00 30 30 30 30  36 34 34 00 30 30 30 30  |....0000644.0000|
00000070  30 30 30 00 30 30 30 31  37 35 30 00 30 30 30 30  |000.0001750.0000|
00000080  30 30 30 30 30 30 30 00  30 30 30 30 30 30 30 30  |0000000.00000000|
00000090  30 30 30 00 30 31 30 33  34 36 00 20 30 00 00 00  |000.010346. 0...|
000000a0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000100  00 75 73 74 61 72 20 20  00 72 6f 6f 74 00 00 00  |.ustar  .root...|
00000110  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000120  00 00 00 00 00 00 00 00  00 74 6d 64 73 00 00 00  |.........tmds...|
00000130  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00002800
```

uid =  30 30 30 30 30 30 30 00 -> 0 in octal representation
gid =  30 30 30 31 37 35 30 00 -> 1000 in octal representation
size = 30 30 30 30 30 30 30 30 30 30 30 00 -> 0 in octal representation
mtime = 30 30 30 30 30 30 30 30 30 30 30 00 -> 0 in octal representation

tmds commented:
> That is what this PR is fixing now. The character 0x30 (not 0x0) to represent zeros. What am I missing?

FormatNumeric calls  FormatOctal

This is where the character 0x30 is meant to be written. If there should be a fix on how value `0` gets handled, it should be in that method (and the fix would be for any octal value written under any format).

carlossanlop commented:
OK need to try again. Last time I tried changing that portion of the code, lots of tests were failing. But time is important.

carlossanlop commented:
Is 0x30 only expected for GNU format? I'm struggling to find this mentioned in the spec(s).

tmds commented:
> Is 0x30 only expected for GNU format? I'm struggling to find this mentioned in the spec(s).

All formats use octal representation for these numeric header fields.

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -1061,6 +1136,14 @@ private static int Checksum(ReadOnlySpan<byte> bytes)
+        private static int FormatAsZeroChars(Span<byte> destination)
+        {
+            Debug.Assert(destination.Length > 1);
+            destination.Slice(0, destination.Length - 1).Fill(0x30); // '0' char
```
Is a casting better than giving the value directly and add a comment?

am11 replied:
If we value the style consistency, then I think, yes. 🙂 

```sh
% git grep "(byte)'" src/libraries/System.Formats.Tar/src
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        RegularFile = (byte)'0',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        HardLink = (byte)'1',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        SymbolicLink = (byte)'2',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        CharacterDevice = (byte)'3',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        BlockDevice = (byte)'4',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        Directory = (byte)'5',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        Fifo = (byte)'6',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        ContiguousFile = (byte)'7',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        ExtendedAttributes = (byte)'x',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        GlobalExtendedAttributes = (byte)'g',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        DirectoryList = (byte)'D',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        LongLink = (byte)'K',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        LongPath = (byte)'L',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        MultiVolume = (byte)'M',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        V7RegularFile = (byte)'\0',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        RenamedOrSymlinked = (byte)'N',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        SparseFile = (byte)'S',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        TapeVolume = (byte)'V',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:            int newlinePos = buffer.IndexOf((byte)'\n');
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:            int spacePos = line.IndexOf((byte)' ');
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:            int equalPos = line.IndexOf((byte)'=');
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:                    span[bytesWritten++] = (byte)' ';
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:                    span[bytesWritten++] = (byte)'=';
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:                    span[bytesWritten++] = (byte)'\n';
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:            checksum += (byte)' ' * 8;
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:            destination[^1] = (byte)' ';
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:            destination[^2] = (byte)'\0';
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:                    destination[i] = (byte)'0';  // Leading zero chars
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:            destination[^1] = (byte)'\0';
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:            destination.Slice(0, copyPos).Fill((byte)'0');
```

carlossanlop replied:
Line deleted anyway.

am11 replied:
If we value the style consistency, then I think, yes. 🙂 

```sh
% git grep "(byte)'" src/libraries/System.Formats.Tar/src
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        RegularFile = (byte)'0',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        HardLink = (byte)'1',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        SymbolicLink = (byte)'2',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        CharacterDevice = (byte)'3',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        BlockDevice = (byte)'4',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        Directory = (byte)'5',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        Fifo = (byte)'6',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        ContiguousFile = (byte)'7',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        ExtendedAttributes = (byte)'x',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        GlobalExtendedAttributes = (byte)'g',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        DirectoryList = (byte)'D',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        LongLink = (byte)'K',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        LongPath = (byte)'L',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        MultiVolume = (byte)'M',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        V7RegularFile = (byte)'\0',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        RenamedOrSymlinked = (byte)'N',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        SparseFile = (byte)'S',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntryType.cs:        TapeVolume = (byte)'V',
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:            int newlinePos = buffer.IndexOf((byte)'\n');
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:            int spacePos = line.IndexOf((byte)' ');
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:            int equalPos = line.IndexOf((byte)'=');
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:                    span[bytesWritten++] = (byte)' ';
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:                    span[bytesWritten++] = (byte)'=';
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:                    span[bytesWritten++] = (byte)'\n';
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:            checksum += (byte)' ' * 8;
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:            destination[^1] = (byte)' ';
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:            destination[^2] = (byte)'\0';
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:                    destination[i] = (byte)'0';  // Leading zero chars
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:            destination[^1] = (byte)'\0';
src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:            destination.Slice(0, copyPos).Fill((byte)'0');
```
(this cast is constant, and roslyn resolves it without baking into the assembly file)

tmds commented:
Octal is base 8. The 0-7 range is written with their ASCII values of 0x30 to 0x37.

tmds commented:
It's the same for all formats.

carlossanlop committed:
Address feedback: set to 0x30 the fields that expect octal and their value is 0 (except the last char). Adjust the new tests to take care of that.

carlossanlop committed:
Update the checksum tests

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -614,25 +673,17 @@ private int WriteCommonFields(Span<byte> buffer, TarEntryType actualEntryType)
 
-            if (_gid > 0)
-            {
-                checksum += FormatNumeric(_gid, buffer.Slice(FieldLocations.Gid, FieldLengths.Gid));
-            }
+            int gid = Math.Max(0, _gid);
+            checksum += FormatNumeric(gid, buffer.Slice(FieldLocations.Gid, FieldLengths.Gid));
 
-            if (_size >= 0)
-            {
-                checksum += FormatNumeric(_size, buffer.Slice(FieldLocations.Size, FieldLengths.Size));
-            }
+            long size = Math.Max(0, _size);
+            checksum += FormatNumeric(size, buffer.Slice(FieldLocations.Size, FieldLengths.Size));
```
@tmds I addressed this for all 4 fields.

tmds replied:
This applied https://github.com/dotnet/runtime/pull/114940#discussion_r2055500705.

I made a different suggestion in the follow up comment https://github.com/dotnet/runtime/pull/114940#discussion_r2055529535.

The latter is to only update gid and uid as in:
```
if (_uid >= 0)
{
    checksum += FormatNumeric(_uid, buffer.Slice(FieldLocations.Uid, FieldLengths.Uid));
}
```

and leave mode and size unchanged.

carlossanlop replied:
Ah sorry. Yes, you're right. I'll bring back the if branches.

What about mode? I know it would be _weird_ to see all zeros in the mode field, but if we see some tar in the wild with such value, we should write it as 0x30 chars too.

tmds replied:
Yes, let's also change `_mode` to be `>= 0`.

carlossanlop replied:
Fixed

tmds commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -1200,7 +1167,10 @@ private static int FormatOctal(long value, Span<byte> destination)
             ulong remaining = (ulong)value;
             Span<byte> digits = stackalloc byte[32]; // longer than any possible octal formatting of a ulong
 
+            digits.Slice(0, digits.Length - 1).Fill(0x30); // Prefill with '0' chars except the last, which should remain 0x0 (null terminator)
```
The `WriteRightAlignedBytesAndGetChecksum` call below takes care of this already.

carlossanlop replied:
Ah, deleting this line. The tests pass still.

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -1117,7 +1167,10 @@ private static int FormatOctal(long value, Span<byte> destination)
             ulong remaining = (ulong)value;
             Span<byte> digits = stackalloc byte[32]; // longer than any possible octal formatting of a ulong
 
+            digits.Slice(0, digits.Length - 1).Fill(0x30); // Prefill with '0' chars except the last, which should remain 0x0 (null terminator)
+
             int i = digits.Length - 1;
+
```
Here's where the bug was: The buffer should've been prefilled with 0x30 chars.

The while loop will be accessed _once_ when `value` is 0, subtracting the `i` iterator _once_ as well.

carlossanlop replied:
Nevermind. This was unnecessary.

carlossanlop committed:
Delete prefill line.

carlossanlop commented:
/backport to release/10.0-preview4

carlossanlop committed:
Bring back the if branches instead of a Math.Max. Make sure to also apply the if check >=0 to mode. Adjust tests to verify mode as well.

carlossanlop commented:
/backport to release/10.0-preview4

carlossanlop committed:
No need for async MemoryStream.WriteAsync (it calls Write internally).

carlossanlop commented:
/backport to release/10.0-preview4

am11 commented:
@tmds, @carlossanlop, in extended attributes name, is it intentional that a tar generated on Windows will have trailing backslash and unix with forwardslash?

https://github.com/dotnet/runtime/blob/ca72cdf4fc66ced0c046f3598ceed6646f48989a/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs#L1177-L1179

carlossanlop commented:
> @tmds, @carlossanlop, in extended attributes name, is it intentional that a tar generated on Windows will have trailing backslash and unix with forwardslash?

Can we discuss that in a separate issue, please? This is unrelated to the bug we're fixing.

am11 commented:
@carlossanlop noticed it in the next method PR is touching, so that's why I've mentioned:

<img width="1508" alt="image" src="https://github.com/user-attachments/assets/35d02024-7f38-42fa-8a07-1ebe180448bc" />

carlossanlop commented:
That's GenerateExtendedAttributeName, which is used by PAX. Not something this PR is touching. Feel free to open a separate issue.

ericstj reviewed (approved):
Looks reasonable and @carlossanlop let me know he's worked with @NikolaMilosavljevic 
 to validate e2e that this fixes the broken deb packages.
 
 Make sure we track any follow up work in an issue or a PR to main.

carlossanlop committed:
Delete temp test.

carlossanlop commented:
@am11 is there a reason for the downvote?

carlossanlop commented:
/backport to release/10.0-preview4

tmds commented:
> is it intentional that a tar generated on Windows will have trailing backslash and unix with forwardslash?

It should all be forward slashes. And as `_name` is expected to only use be using forward slashes as directory separators, `GenerateExtendedAttributeName` shouldn't use `Path.GetDirectoryName`/`Path.GetFileName` as on Windows it would also treat backward slashes as directory separators.

carlossanlop commented:
/ba-g failures are unrelated to tar
